File: engine/src/mcts.cpp
// engine/src/mcts.cpp
// YichiAlpha — MCTS implementation (C++)

#include "mcts.h"
#include "rules.h"

#include <algorithm>
#include <cassert>
#include <cmath>
#include <random>
#include <stdexcept>

namespace yichi {

// ---------------------------------------------------------------------------
// MCTSNode
// ---------------------------------------------------------------------------
MCTSNode::MCTSNode(Board s, MCTSNode* p, std::pair<int,int> m, float pr)
    : state(std::move(s)), parent(p), move(m), prior(pr) {}

float MCTSNode::ucb_score(float c_puct) const {
    if (parent == nullptr) return 0.0f;
    float u = c_puct * prior *
              (std::sqrt(static_cast<float>(parent->N.load())) /
               (1.0f + static_cast<float>(N.load())));
    // Q is stored from "this node's player's perspective".
    // Parent's player is the opponent, so parent's value = -Q.
    return -Q + u;
}

MCTSNode* MCTSNode::best_child(float c_puct) const {
    MCTSNode* best = nullptr;
    float best_score = -1e9f;
    for (auto& c : children) {
        float s = c->ucb_score(c_puct);
        if (s > best_score) {
            best_score = s;
            best = c.get();
        }
    }
    return best;
}

// ---------------------------------------------------------------------------
// MCTS
// ---------------------------------------------------------------------------
MCTS::MCTS(YichiNet model, float c_puct, int n_simulations,
           float dirichlet_alpha, float dirichlet_epsilon, int batch_size,
           torch::Device device)
    : model_(std::move(model)),
      c_puct_(c_puct),
      n_simulations_(n_simulations),
      dirichlet_alpha_(dirichlet_alpha),
      dirichlet_epsilon_(dirichlet_epsilon),
      batch_size_(batch_size),
      board_size_(0),
      device_(device) {
    // board_size_ will be set per search from root_state
    // Move model to the target device (CPU or CUDA)
    model_->to(device_);
}

torch::Tensor board_to_tensor(const Board& board, torch::Device device) {
    const int N = board.board_size();
    const int p = board.current_player();
    const int opp = (p == 1) ? 2 : 1;

    // Build on CPU first (faster for small tensor initialization)
    auto options = torch::TensorOptions().dtype(torch::kFloat32).device(torch::kCPU);
    torch::Tensor planes = torch::zeros({1, 11, N, N}, options);
    auto accessor = planes.accessor<float, 4>();
    for (int r = 0; r < N; ++r) {
        for (int c = 0; c < N; ++c) {
            int t = static_cast<int>(board.at(r, c));
            int h = board.hp_at(r, c);
            if (t == p) {
                accessor[0][0][r][c] = 1.0f;
                if (h == 1)       accessor[0][3][r][c] = 1.0f;
                else if (h == 2)  accessor[0][4][r][c] = 1.0f;
                else if (h >= 3)  accessor[0][5][r][c] = 1.0f;
            } else if (t == opp) {
                accessor[0][1][r][c] = 1.0f;
                if (h == 1)       accessor[0][6][r][c] = 1.0f;
                else if (h == 2)  accessor[0][7][r][c] = 1.0f;
                else if (h >= 3)  accessor[0][8][r][c] = 1.0f;
            } else if (t == 3) {  // BLOCK
                accessor[0][2][r][c] = 1.0f;
            }
        }
    }
    // Legal moves mask
    auto legal = board.legal_moves();
    for (auto& mv : legal) {
        accessor[0][9][mv.first][mv.second] = 1.0f;
    }
    // Side to move (constant 1 after perspective alignment)
    for (int r = 0; r < N; ++r)
        for (int c = 0; c < N; ++c)
            accessor[0][10][r][c] = 1.0f;

    // Move to target device if not CPU
    if (device.is_cuda()) {
        planes = planes.to(device);
    }
    return planes;
}

void MCTS::expand_node(MCTSNode* node, const std::vector<float>& policy, bool add_noise) {
    auto legal = node->state.legal_moves();
    const int N = node->state.board_size();

    std::vector<float> priors;
    priors.reserve(legal.size());
    for (auto& mv : legal) {
        int idx = mv.first * N + mv.second;
        priors.push_back(policy[idx]);
    }

    if (add_noise && !legal.empty()) {
        std::gamma_distribution<float> gamma(dirichlet_alpha_, 1.0f);
        std::mt19937 rng(std::random_device{}());
        std::vector<float> noise(legal.size());
        float noise_sum = 0.0f;
        for (size_t i = 0; i < legal.size(); ++i) {
            noise[i] = gamma(rng);
            noise_sum += noise[i];
        }
        if (noise_sum > 0) {
            for (size_t i = 0; i < legal.size(); ++i) {
                priors[i] = (1.0f - dirichlet_epsilon_) * priors[i] +
                            dirichlet_epsilon_ * (noise[i] / noise_sum);
            }
        }
    }

    // Renormalize
    float sum = 0.0f;
    for (float p : priors) sum += p;
    if (sum > 0) {
        for (float& p : priors) p /= sum;
    } else {
        std::fill(priors.begin(), priors.end(), 1.0f / priors.size());
    }

    for (size_t i = 0; i < legal.size(); ++i) {
        Board child = node->state.clone();
        Rules::apply_move(child, legal[i].first, legal[i].second);
        node->children.push_back(std::make_unique<MCTSNode>(
            std::move(child), node, legal[i], priors[i]));
    }
    node->is_expanded = true;
}

void MCTS::backup(MCTSNode* node, float value) {
    while (node != nullptr) {
        node->N.fetch_add(1);
        float old_w = node->W.load();
        while (!node->W.compare_exchange_weak(old_w, old_w + value)) {}
        int n = node->N.load();
        node->Q = node->W.load() / static_cast<float>(n);
        value = -value;  // flip perspective
        node = node->parent;
    }
}

void MCTS::evaluate_batch(bool root_add_noise) {
    int n = static_cast<int>(eval_queue_.size());
    if (n == 0) return;

    // Build batched input (on CPU, then move to device_ once)
    std::vector<torch::Tensor> inputs;
    inputs.reserve(n);
    for (int i = 0; i < n; ++i) {
        inputs.push_back(board_to_tensor(eval_queue_[i]->state, torch::kCPU));
    }
    torch::Tensor batch = torch::cat(inputs, /*dim=*/0);
    batch = batch.to(device_);

    // Forward
    auto [logits, value] = model_->forward(batch);
    auto policy = torch::softmax(logits, /*dim=*/1);

    // Move outputs to CPU for accessor (works regardless of device)
    policy = policy.to(torch::kCPU);
    value = value.to(torch::kCPU);

    auto policy_acc = policy.accessor<float, 2>();
    auto value_acc = value.accessor<float, 2>();
    int N = eval_queue_[0]->state.board_size();

    for (int i = 0; i < n; ++i) {
        std::vector<float> p(N * N + 1);
        for (int j = 0; j < N * N + 1; ++j) p[j] = policy_acc[i][j];

        bool add_noise = (eval_queue_[i]->parent == nullptr) && root_add_noise;
        expand_node(eval_queue_[i], p, add_noise);
        backup(eval_queue_[i], value_acc[i][0]);
    }
    eval_queue_.clear();
}

void MCTS::select_and_expand(MCTSNode* root, bool root_add_noise) {
    MCTSNode* node = root;
    while (node->is_expanded && !node->state.is_terminal()) {
        node = node->best_child(c_puct_);
    }

    if (node->state.is_terminal()) {
        float v = node->state.terminal_reward_for_current_player();
        backup(node, v);
        return;
    }

    eval_queue_.push_back(node);
    if (static_cast<int>(eval_queue_.size()) >= batch_size_) {
        evaluate_batch(root_add_noise);
    }
}

std::unique_ptr<MCTSNode> MCTS::search(const Board& root_state, bool add_noise) {
    board_size_ = root_state.board_size();
    auto root = std::make_unique<MCTSNode>(root_state.clone(), nullptr,
                                            std::make_pair(-1, -1), 0.0f);

    // Evaluate root
    auto input = board_to_tensor(root->state, device_);
    auto [logits, value] = model_->forward(input);
    auto policy = torch::softmax(logits, /*dim=*/1);
    // Move policy back to CPU for accessor (always works regardless of device)
    policy = policy.to(torch::kCPU);
    value = value.to(torch::kCPU);
    auto policy_acc = policy.accessor<float, 2>();
    int N = board_size_;
    std::vector<float> p(N * N + 1);
    for (int j = 0; j < N * N + 1; ++j) p[j] = policy_acc[0][j];
    expand_node(root.get(), p, add_noise);

    if (root->state.is_terminal()) return root;

    for (int i = 0; i < n_simulations_; ++i) {
        select_and_expand(root.get(), add_noise);
    }
    // Flush any remaining
    evaluate_batch(add_noise);
    return root;
}

std::vector<float> MCTS::get_action_distribution(
    const MCTSNode& root, float temperature) const {

    const int N = root.state.board_size();
    std::vector<float> visits(N * N + 1, 0.0f);

    for (const auto& c : root.children) {
        int idx = c->move.first * N + c->move.second;
        visits[idx] = static_cast<float>(c->N.load());
    }

    if (temperature == 0.0f) {
        int max_idx = 0;
        float max_v = -1.0f;
        for (int i = 0; i < N * N + 1; ++i) {
            if (visits[i] > max_v) {
                max_v = visits[i];
                max_idx = i;
            }
        }
        std::vector<float> result(N * N + 1, 0.0f);
        result[max_idx] = 1.0f;
        return result;
    }

    float sum = 0.0f;
    for (float& v : visits) {
        v = std::pow(v, 1.0f / temperature);
        sum += v;
    }
    if (sum > 0) {
        for (float& v : visits) v /= sum;
    } else {
        std::fill(visits.begin(), visits.end(), 1.0f / visits.size());
    }
    return visits;
}

std::pair<int,int> MCTS::select_action(
    const MCTSNode& root, float temperature, unsigned long seed) const {

    auto visits = get_action_distribution(root, temperature);
    const int N = root.state.board_size();

    std::mt19937 rng(seed ? seed : std::random_device{}());
    std::discrete_distribution<int> dist(visits.begin(), visits.end());
    int idx = dist(rng);

    if (idx >= N * N) {
        // Pass fallback
        auto legal = root.state.legal_moves();
        return legal.empty() ? std::make_pair(-1, -1) : legal[0];
    }
    return {idx / N, idx % N};
}

}  // namespace yichi

File: python/mcts.py
"""
YichiAlpha — MCTS with Neural Network guidance
================================================
Monte Carlo Tree Search adapted for 异吃棋 chain reactions.

Key design (see docs/04_mcts_adaptation.md):
  - Transition = (move + full chain refresh until stable)
  - value = current-player perspective
  - backup with perspective flip
  - Dirichlet noise at root for exploration
"""
from __future__ import annotations

import math
import numpy as np
import torch
import torch.nn.functional as F
from typing import List, Tuple, Optional, Dict
from dataclasses import dataclass, field

from game import (
    GameState, GameConfig, X, O, BLOCK, EMPTY,
    state_to_tensor, move_to_index, index_to_move,
    N_FEATURE_CHANNELS,
)
from model import YichiNet


# ---------------------------------------------------------------------------
# MCTS Node
# ---------------------------------------------------------------------------
@dataclass
class MCTSNode:
    """A node in the MCTS search tree."""
    state: GameState
    parent: Optional['MCTSNode'] = None
    move: Optional[Tuple[int, int]] = None    # move that led here (None for root)
    prior: float = 0.0                         # P(s, a) from network
    children: Dict[Tuple[int, int], 'MCTSNode'] = field(default_factory=dict)
    N: int = 0
    W: float = 0.0
    Q: float = 0.0
    is_expanded: bool = False

    def ucb_score(self, c_puct: float) -> float:
        """PUCT: Q + c_puct * P * sqrt(N_parent) / (1 + N)

        Note: self.Q is stored from "the player to move at this node's
        perspective". When the parent is choosing among children, the
        parent's player is the OPPONENT of this node's player, so the
        parent's value for moving to this child is -self.Q.
        """
        if self.parent is None:
            return 0.0
        u = c_puct * self.prior * (math.sqrt(self.parent.N) / (1 + self.N))
        return -self.Q + u  # negate: parent wants to MAXIMIZE its own value

    def best_child(self, c_puct: float) -> 'MCTSNode':
        return max(self.children.values(), key=lambda c: c.ucb_score(c_puct))


# ---------------------------------------------------------------------------
# MCTS Searcher
# ---------------------------------------------------------------------------
class MCTS:
    """AlphaZero-style MCTS with NN guidance.

    Parameters
    ----------
    model : YichiNet
        Policy-Value network.
    c_puct : float
        Exploration constant. Default 1.5 (higher than AlphaZero's 1.0 because
        异吃棋 value signal is weaker).
    n_simulations : int
        Number of MCTS simulations per move.
    dirichlet_alpha : float
        Dirichlet noise concentration. 0.3 for ~37 actions (similar to chess).
    dirichlet_epsilon : float
        Noise mixing weight (0.25 standard).
    device : str
        'cpu' or 'cuda'.
    """

    def __init__(self, model: YichiNet, c_puct: float = 1.5,
                 n_simulations: int = 200,
                 dirichlet_alpha: float = 0.3,
                 dirichlet_epsilon: float = 0.25,
                 device: str = 'cpu'):
        self.model = model
        self.c_puct = c_puct
        self.n_simulations = n_simulations
        self.dirichlet_alpha = dirichlet_alpha
        self.dirichlet_epsilon = dirichlet_epsilon
        self.device = torch.device(device)
        self.model.to(self.device)

    # ----------------------------------------------------------------
    # Public API
    # ----------------------------------------------------------------
    def search(self, root_state: GameState, add_noise: bool = True) -> MCTSNode:
        """Run MCTS from root_state, return the root node (with stats)."""
        root = MCTSNode(state=root_state.clone())

        # Evaluate root
        policy, value = self._evaluate(root.state)
        self._expand(root, policy, add_noise=add_noise)

        # If root is terminal (rare — only if board is full), return immediately
        if root.state.is_terminal():
            return root

        for _ in range(self.n_simulations):
            node = self._select(root)
            v = self._evaluate_and_expand(node)
            self._backup(node, v)

        return root

    def get_action_distribution(self, root: MCTSNode,
                                 temperature: float = 1.0) -> np.ndarray:
        """Return policy target π ~ visits^(1/T) over action space (N*N+1,)."""
        n = root.state.board_size
        visits = np.zeros(n * n + 1, dtype=np.float32)
        for move, child in root.children.items():
            idx = move_to_index(move, n)
            visits[idx] = child.N

        if temperature == 0:
            # Greedy
            result = np.zeros_like(visits)
            result[visits.argmax()] = 1.0
            return result

        visits = visits ** (1.0 / temperature)
        total = visits.sum()
        if total == 0:
            # Fallback: uniform
            return np.ones_like(visits) / len(visits)
        return visits / total

    def select_action(self, root: MCTSNode, temperature: float = 1.0,
                       rng: Optional[np.random.Generator] = None) -> Tuple[int, int]:
        """Sample an action from the MCTS visit distribution."""
        pi = self.get_action_distribution(root, temperature)
        if rng is None:
            rng = np.random.default_rng()
        idx = rng.choice(len(pi), p=pi)
        # Pass action (idx == N*N) — extremely rare, only when no legal moves
        if idx == root.state.board_size ** 2:
            # Shouldn't happen because apply_move requires legal move
            # Fallback: pick first legal
            legal = root.state.legal_moves()
            return legal[0] if legal else (0, 0)
        return index_to_move(idx, root.state.board_size)

    # ----------------------------------------------------------------
    # Internal: selection, evaluation, expansion, backup
    # ----------------------------------------------------------------
    def _select(self, root: MCTSNode) -> MCTSNode:
        """Traverse tree by PUCT until leaf."""
        node = root
        while node.is_expanded and not node.state.is_terminal():
            node = node.best_child(self.c_puct)
        return node

    def _evaluate_and_expand(self, node: MCTSNode) -> float:
        """Evaluate leaf with network, expand, return value (node's perspective)."""
        if node.state.is_terminal():
            return node.state.terminal_reward_for_current_player()

        policy, value = self._evaluate(node.state)
        self._expand(node, policy, add_noise=False)
        return float(value)

    def _evaluate(self, state: GameState) -> Tuple[np.ndarray, float]:
        """Run network on state, return (policy_probs, value_scalar)."""
        x = state_to_tensor(state)
        x = torch.from_numpy(x).unsqueeze(0).to(self.device)
        self.model.eval()
        with torch.no_grad():
            logits, value = self.model(x)
            policy = F.softmax(logits, dim=-1)
        return policy.squeeze(0).cpu().numpy(), float(value.squeeze().item())

    def _expand(self, node: MCTSNode, policy: np.ndarray, add_noise: bool):
        """Create child nodes for all legal moves."""
        legal_moves = node.state.legal_moves()
        n = node.state.board_size

        if add_noise and len(legal_moves) > 0:
            noise = np.random.dirichlet(
                [self.dirichlet_alpha] * len(legal_moves)
            )
        else:
            noise = np.zeros(len(legal_moves))

        # Normalize priors over legal moves only
        priors = np.array([policy[move_to_index(m, n)] for m in legal_moves])
        if add_noise:
            priors = (1 - self.dirichlet_epsilon) * priors + self.dirichlet_epsilon * noise
        # Renormalize
        total = priors.sum()
        if total > 0:
            priors = priors / total
        else:
            # Uniform fallback
            priors = np.ones(len(legal_moves)) / len(legal_moves)

        for move, p in zip(legal_moves, priors):
            child_state = node.state.clone()
            child_state.apply_move(move)
            node.children[move] = MCTSNode(
                state=child_state, parent=node, move=move, prior=float(p)
            )
        node.is_expanded = True

    def _backup(self, node: MCTSNode, value: float):
        """Backup value with perspective flip."""
        while node is not None:
            node.N += 1
            node.W += value
            node.Q = node.W / node.N
            value = -value  # flip perspective
            node = node.parent


# ---------------------------------------------------------------------------
# Self-test
# ---------------------------------------------------------------------------
if __name__ == '__main__':
    import time

    print("=== MCTS Self-test ===")
    model = YichiNet(board_size=6)
    mcts = MCTS(model, n_simulations=50, device='cpu')

    state = GameState.initial()
    print("Initial state:")
    print(state)
    print()

    t0 = time.time()
    root = mcts.search(state, add_noise=True)
    elapsed = time.time() - t0
    print(f"Search done in {elapsed:.2f}s ({mcts.n_simulations} sims)")

    pi = mcts.get_action_distribution(root, temperature=1.0)
    print(f"Action distribution sum: {pi.sum():.4f}")
    print(f"Top 5 actions:")
    n = state.board_size
    top5 = np.argsort(pi)[::-1][:5]
    for idx in top5:
        if idx < n * n:
            r, c = index_to_move(idx, n)
            print(f"  ({r},{c}): {pi[idx]:.3f}")
        else:
            print(f"  pass: {pi[idx]:.3f}")

    # Pick an action
    move = mcts.select_action(root, temperature=1.0)
    print(f"\nSelected move: {move}")
    state.apply_move(move)
    print("After move:")
    print(state)
